File: php/phpfind/src/phpfind/ArgTokenizer.php
<?php

declare(strict_types=1);

namespace phpfind;

readonly class ArgTokenizer {
    private array $bool_map;
    private array $str_map;
    private array $int_map;

    public function __construct(array $bool_map, array $str_map, array $int_map)
    {
        $this->bool_map = $bool_map;
        $this->str_map = $str_map;
        $this->int_map = $int_map;
    }

    /**
     * @param string[] $args
     * @return array
     * @throws FindException
     */
    public function tokenize_args(array $args): array
    {
        $arg_tokens = [];
        while (count($args) > 0) {
            $arg = array_shift($args);
            if ($arg[0] == '-') {
                $arg_names = array();
                if (strlen($arg) > 1) {
                    if ($arg[1] == '-') {
                        # it's a long arg
                        if (strlen($arg) > 2) {
                            $arg = substr($arg, 2);
                            if (str_contains($arg, '=')) {
                                $elems = explode('=', $arg);
                                $arg = $elems[0];
                                $arg_value = $elems[1];
                                array_unshift($args, $arg_value);
                            }
                            $arg_names[] = $arg;
                        } else {
                            throw new FindException("Invalid option: $arg");
                        }
                    } else {
                        # it's a short arg, with possibly multiple args
                        $arg = substr($arg, 1);
                        foreach (str_split($arg) as $c) {
                            if (array_key_exists($c, $this->bool_map)) {
                                $arg_names[] = $this->bool_map[$c];
                            } elseif (array_key_exists($c, $this->str_map)) {
                                $arg_names[] = $this->str_map[$c];
                            } elseif (array_key_exists($c, $this->int_map)) {
                                $arg_names[] = $this->int_map[$c];
                            } else {
                                throw new FindException("Invalid option: $c");
                            }
                        }
                    }
                } else {
                    throw new FindException("Invalid option: $arg");
                }

                foreach ($arg_names as $arg_name) {
                    if (array_key_exists($arg_name, $this->bool_map)) {
                        $arg_tokens[] = new ArgToken($arg_name, ArgTokenType::Bool, true);
                        if (in_array($arg_name, array("help", "version"))) {
                            return $arg_tokens;
                        }
                    } elseif (array_key_exists($arg_name, $this->str_map)
                        || array_key_exists($arg_name, $this->int_map)) {
                        if (count($args) > 0) {
                            $val = array_shift($args);
                            if (array_key_exists($arg_name, $this->str_map)) {
                                $arg_tokens[] = new ArgToken($arg_name, ArgTokenType::Str, $val);
                            } elseif (array_key_exists($arg_name, $this->int_map)) {
                                $arg_tokens[] = new ArgToken($arg_name, ArgTokenType::Int, intval($val));
                            }
                        } else {
                            throw new FindException("Missing value for $arg");
                        }
                    } else {
                        throw new FindException("Invalid option: $arg");
                    }
                }
            } else {
                $arg_tokens[] = new ArgToken('path', ArgTokenType::Str, $arg);
            }
        }
        return $arg_tokens;
    }

    /**
     * @param array $arg_map
     * @return array
     * @throws FindException
     */
    public function tokenize_arg_map(array $arg_map): array
    {
        $arg_names = array_keys($arg_map);
        # keys are sorted so that output is consistent across all versions
        sort($arg_names);
        $arg_tokens = [];
        foreach ($arg_names as $arg_name) {
            $arg_value = $arg_map[$arg_name];
            if (array_key_exists($arg_name, $this->bool_map)) {
                $long_arg = $this->bool_map[$arg_name];
                if (is_bool($arg_value)) {
                    $arg_tokens[] = new ArgToken($long_arg, ArgTokenType::Bool, $arg_value);
                    if (in_array($long_arg, array("help", "version"))) {
                        return $arg_tokens;
                    }
                } else {
                    throw new FindException("Invalid value for option: $arg_name");
                }
            } elseif (array_key_exists($arg_name, $this->str_map)) {
                $long_arg = $this->str_map[$arg_name];
                if (is_array($arg_value)) {
                    foreach ($arg_value as $val) {
                        if (is_string($val)) {
                            $arg_tokens[] = new ArgToken($long_arg, ArgTokenType::Str, $val);
                        } else {
                            throw new FindException("Invalid value for option: $arg_name");
                        }
                    }
                } elseif (is_string($arg_value)) {
                    $arg_tokens[] = new ArgToken($long_arg, ArgTokenType::Str, $arg_value);
                } else {
                    throw new FindException("Invalid value for option: $arg_name");
                }
            } elseif (array_key_exists($arg_name, $this->int_map)) {
                $long_arg = $this->int_map[$arg_name];
                if (is_int($arg_value)) {
                    $arg_tokens[] = new ArgToken($long_arg, ArgTokenType::Int, $arg_value);
                } elseif (is_string($arg_value)) {
                    $arg_tokens[] = new ArgToken($long_arg, ArgTokenType::Int, intval($arg_value));
                } else {
                    throw new FindException("Invalid value for option: $arg_name");
                }
            } else {
                throw new FindException("Invalid option: $arg_name");
            }
        }
        return $arg_tokens;
    }

    /**
     * @param string $json
     * @return array
     * @throws FindException
     */
    public function tokenize_json(string $json): array
    {
        if (trim($json) === '') {
            return [];
        }
        try {
            $json_obj = (array)json_decode(trim($json), true, 512, JSON_THROW_ON_ERROR);
            return $this->tokenize_arg_map($json_obj);
        } catch (\JsonException $e) {
            throw new FindException($e->getMessage());
        }
    }

    /**
     * @param string $file_path
     * @return array
     * @throws FindException
     */
    public function tokenize_file(string $file_path): array
    {
        $expanded_path = FileUtil::expand_path($file_path);
        if (!file_exists($expanded_path)) {
            throw new FindException('Settings file not found: ' . $file_path);
        }
        if (!str_ends_with($expanded_path, '.json')) {
            throw new FindException('Invalid settings file (must be JSON): ' . $file_path);
        }
        $json = file_get_contents($expanded_path);
        return $this->tokenize_json($json);
    }
}

File: cpp/cppfind/src/ArgTokenizer.cpp
#include "ArgTokenizer.h"

#include <rapidjson/filereadstream.h>

#include "FileUtil.h"
#include "FindException.h"
#include "StringUtil.h"


namespace cppfind {
    ArgTokenizer::ArgTokenizer() = default;

    ArgTokenizer::ArgTokenizer(const std::unordered_map<std::string, std::string> &bool_map,
        const std::unordered_map<std::string, std::string> &str_map,
        const std::unordered_map<std::string, std::string> &int_map,
        const std::unordered_map<std::string, std::string> &long_map)
    : m_bool_map{bool_map},  m_str_map{str_map},  m_int_map{int_map},  m_long_map{long_map} {
    }

    std::vector<ArgToken> ArgTokenizer::tokenize_args(int &argc, char **argv) const {
        std::vector<ArgToken> arg_tokens{};
        std::deque<std::string> arg_deque;
        int i;

        for (i=1; i < argc; ++i) {
            arg_deque.emplace_back(argv[i]);
        }

        std::string next_arg;
        while (!arg_deque.empty()) {
            next_arg = arg_deque.front();
            arg_deque.pop_front();

            if (next_arg.rfind('-', 0) == 0) {
                std::deque<std::string> next_arg_deque;
                std::string next_val;
                if (next_arg.rfind("--", 0) == 0) {
                    if (next_arg.size() < 3) {
                        std::string msg{"Invalid option: " + next_arg};
                        msg.append(next_arg);
                        throw FindException(msg);
                    }
                    next_arg = next_arg.substr(2);
                    std::vector<std::string> parts = StringUtil::split_string(next_arg, '=');
                    if (!parts.empty()) {
                        next_arg = parts[0];
                    }
                    if (parts.size() > 1) {
                        next_val = parts[1];
                    }
                    next_arg_deque.emplace_back(next_arg);

                } else if (next_arg.size() > 1) {
                    next_arg = next_arg.substr(1);
                    for (const char & c : next_arg) {
                        std::string cs(1, c);
                        next_arg_deque.emplace_back(cs);
                    }

                } else {
                    std::string msg{"Invalid option: " + next_arg};
                    msg.append(next_arg);
                    throw FindException(msg);
                }

                while (!next_arg_deque.empty()) {
                    next_arg = next_arg_deque.front();
                    next_arg_deque.pop_front();

                    if (m_bool_map.contains(next_arg)) {
                        std::string long_arg = m_bool_map.at(next_arg);
                        arg_tokens.emplace_back(long_arg, ARG_TOKEN_TYPE_BOOL, true);
                    } else if (m_str_map.contains(next_arg)
                        || m_int_map.contains(next_arg)
                        || m_long_map.contains(next_arg)
                        || next_arg == "settings-file") {

                        if (next_val.empty()) {
                            if (arg_deque.empty()) {
                                std::string msg{"Missing value for option "};
                                msg.append(next_arg);
                                throw FindException(msg);
                            }
                            next_val = std::string(arg_deque.front());
                            arg_deque.pop_front();
                        }
                        if (m_str_map.contains(next_arg)) {
                            std::string long_arg = m_str_map.at(next_arg);
                            arg_tokens.emplace_back(long_arg, ARG_TOKEN_TYPE_STR, next_val);
                        } else if (m_int_map.contains(next_arg)) {
                            std::string long_arg = m_int_map.at(next_arg);
                            const int int_val = std::stoi(next_val);
                            arg_tokens.emplace_back(long_arg, ARG_TOKEN_TYPE_INT, int_val);
                        } else if (m_long_map.contains(next_arg)) {
                            std::string long_arg = m_long_map.at(next_arg);
                            const long long_val = std::stol(next_val);
                            arg_tokens.emplace_back(long_arg, ARG_TOKEN_TYPE_LONG, long_val);
                        } else {
                            // settings-file
                            arg_tokens.emplace_back(next_arg, ARG_TOKEN_TYPE_STR, next_val);
                        }
                    } else {
                        std::string msg{"Invalid option: "};
                        msg.append(next_arg);
                        throw FindException(msg);
                    }
                }
            } else {
                arg_tokens.emplace_back("path", ARG_TOKEN_TYPE_STR, next_arg);
            }
        }
        return arg_tokens;
    }

    std::vector<ArgToken> ArgTokenizer::tokenize_document(rapidjson::Document& document) const {
        assert(document.IsObject());

        std::vector<ArgToken> arg_tokens{};

        // Get the property names
        std::vector<std::string> names;
        for (rapidjson::Value::ConstMemberIterator it=document.MemberBegin(); it != document.MemberEnd(); ++it) {
            std::string name = it->name.GetString();
            names.push_back(name);
        }

        // Sort the names - produces consistent behavior across versions
        std::ranges::sort(names);

        // Iterate through names and values, validating values and applying to settings
        for (rapidjson::Value::ConstMemberIterator it=document.MemberBegin(); it != document.MemberEnd(); ++it) {
            if (std::string name = it->name.GetString(); m_bool_map.contains(name)) {
                if (it->value.IsBool()) {
                    arg_tokens.emplace_back(name, ARG_TOKEN_TYPE_BOOL, it->value.GetBool());
                } else {
                    std::string msg{"Invalid value for option: " + name};
                    throw FindException(msg);
                }
            } else if (m_str_map.contains(name) || name == "settings-file") {
                if (it->value.IsString()) {
                    auto s = std::string(it->value.GetString());
                    arg_tokens.emplace_back(name, ARG_TOKEN_TYPE_STR, s);
                } else if (it->value.IsArray()) {
                    const auto& arr = it->value.GetArray();
                    for (rapidjson::SizeType i = 0; i < arr.Size(); ++i) {
                        if (arr[i].IsString()) {
                            auto s = std::string(arr[i].GetString());
                            arg_tokens.emplace_back(name, ARG_TOKEN_TYPE_STR, s);
                        } else {
                            std::string msg{"Invalid value for option: " + name};
                            throw FindException(msg);
                        }
                    }
                } else {
                    std::string msg{"Invalid value for option: " + name};
                    throw FindException(msg);
                }
            } else if (m_int_map.contains(name)) {
                if (it->value.IsInt()) {
                    arg_tokens.emplace_back(name, ARG_TOKEN_TYPE_INT, it->value.GetInt());
                } else if (it->value.IsUint()) {
                    arg_tokens.emplace_back(name, ARG_TOKEN_TYPE_INT, it->value.GetUint());
                } else if (it->value.IsInt64()) {
                    const auto l = it->value.GetInt64();
                    const int i = 0 + l;
                    arg_tokens.emplace_back(name, ARG_TOKEN_TYPE_INT, i);
                } else if (it->value.IsUint64()) {
                    const auto l = it->value.GetUint64();
                    const int i = 0 + l;
                    arg_tokens.emplace_back(name, ARG_TOKEN_TYPE_INT, i);
                } else {
                    std::string msg{"Invalid value for option: " + name};
                    throw FindException(msg);
                }
            } else if (m_long_map.contains(name)) {
                if (it->value.IsInt()) {
                    const auto i = it->value.GetInt();
                    const long l = 0 + i;
                    arg_tokens.emplace_back(name, ARG_TOKEN_TYPE_LONG, l);
                } else if (it->value.IsUint()) {
                    const auto i = it->value.GetUint();
                    const long l = 0 + i;
                    arg_tokens.emplace_back(name, ARG_TOKEN_TYPE_LONG, l);
                } else if (it->value.IsInt64()) {
                    arg_tokens.emplace_back(name, ARG_TOKEN_TYPE_LONG, it->value.GetInt64());
                } else if (it->value.IsUint64()) {
                    arg_tokens.emplace_back(name, ARG_TOKEN_TYPE_LONG, it->value.GetUint64());
                } else {
                    std::string msg{"Invalid value for option: " + name};
                    throw FindException(msg);
                }
            } else {
                // Shouldn't be able to get here since we already checked names
                std::string msg{"Invalid option: " + name};
                throw FindException(msg);
            }
        }
        return arg_tokens;
    }

    std::vector<ArgToken> ArgTokenizer::tokenize_json(const std::string_view json) const {
        rapidjson::Document document;
        document.Parse(std::string{json}.c_str());
        return tokenize_document(document);
    }

    std::vector<ArgToken> ArgTokenizer::tokenize_file(const std::filesystem::path& file_path) const {
        std::filesystem::path expanded_path = FileUtil::expand_path(file_path);
        if (!std::filesystem::exists(expanded_path)) {
            std::string msg{"Settings file not found: "};
            msg.append(file_path);
            throw FindException(msg);
        }

        if (file_path.extension() != ".json") {
            std::string msg{"Invalid settings file (must be JSON): "};
            msg.append(file_path);
            throw FindException(msg);
        }

        const uint64_t file_size = std::filesystem::file_size(expanded_path);
        // ~1MB, an arbitrary limit, but at least a limit
        assert(file_size <= 1024000);

        FILE *fp = fopen(expanded_path.c_str(), "r");

        char readBuffer[file_size];
        rapidjson::FileReadStream is(fp, readBuffer, sizeof(readBuffer));

        rapidjson::Document document;
        document.ParseStream(is);
        fclose(fp);

        return tokenize_document(document);
    }

}
